https://github.com/cupy/cupy/pull/8319
From 39382bab4c85b92ac2a33e603471faed63c39003 Mon Sep 17 00:00:00 2001
From: Nakano Azusa <31366253+Azusachan@users.noreply.github.com>
Date: Fri, 3 May 2024 01:00:51 +0800
Subject: [PATCH 1/2] Support Build on ROCM >= 6.0.0
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

ROCm 6 introduced various changes on its API. In particular,
* Removal of gcnarch from hipDeviceProp_t structure
* Renaming of ‘memoryType’ in hipPointerAttribute_t structure to ‘type’

This patch allows cupy to be built on this version.
---
 cupy_backends/cuda/api/_runtime_typedef.pxi |  8 +++++++-
 cupy_backends/cuda/api/runtime.pyx          | 11 +++++++++--
 cupy_backends/hip/cupy_hip_runtime.h        |  6 +++---
 install/cupy_builder/install_build.py       |  5 +++++
 4 files changed, 24 insertions(+), 6 deletions(-)

diff --git a/cupy_backends/cuda/api/_runtime_typedef.pxi b/cupy_backends/cuda/api/_runtime_typedef.pxi
index e9e243ee44b..053a6eccd38 100644
--- a/cupy_backends/cuda/api/_runtime_typedef.pxi
+++ b/cupy_backends/cuda/api/_runtime_typedef.pxi
@@ -138,12 +138,18 @@ cdef extern from *:
             int device
             void* devicePointer
             void* hostPointer
-    ELIF CUPY_HIP_VERSION > 0:
+    ELIF 0 < CUPY_HIP_VERSION < 600:
         ctypedef struct _PointerAttributes 'cudaPointerAttributes':
             int memoryType
             int device
             void* devicePointer
             void* hostPointer
+    ELIF CUPY_HIP_VERSION >= 600:
+        ctypedef struct _PointerAttributes 'cudaPointerAttributes':
+            int type
+            int device
+            void* devicePointer
+            void* hostPointer
     ELSE:
         ctypedef struct _PointerAttributes 'cudaPointerAttributes':
             pass  # for RTD
diff --git a/cupy_backends/cuda/api/runtime.pyx b/cupy_backends/cuda/api/runtime.pyx
index 58cb5af54fe..d2d0e4585c0 100644
--- a/cupy_backends/cuda/api/runtime.pyx
+++ b/cupy_backends/cuda/api/runtime.pyx
@@ -318,7 +318,6 @@ cpdef getDeviceProperties(int device):
         properties['clockInstructionRate'] = props.clockInstructionRate
         properties['maxSharedMemoryPerMultiProcessor'] = (
             props.maxSharedMemoryPerMultiProcessor)
-        properties['gcnArch'] = props.gcnArch
         properties['hdpMemFlushCntl'] = <intptr_t>(props.hdpMemFlushCntl)
         properties['hdpRegFlushCntl'] = <intptr_t>(props.hdpRegFlushCntl)
         properties['memPitch'] = props.memPitch
@@ -351,6 +350,8 @@ cpdef getDeviceProperties(int device):
         arch['has3dGrid'] = props.arch.has3dGrid
         arch['hasDynamicParallelism'] = props.arch.hasDynamicParallelism
         properties['arch'] = arch
+    IF CUPY_HIP_VERSION < 600: # removed in HIP 6.0.0
+        properties['gcnArch'] = props.gcnArch
     IF CUPY_HIP_VERSION >= 310:
         properties['gcnArchName'] = props.gcnArchName
         properties['asicRevision'] = props.asicRevision
@@ -720,12 +721,18 @@ cpdef PointerAttributes pointerGetAttributes(intptr_t ptr):
             <intptr_t>attrs.devicePointer,
             <intptr_t>attrs.hostPointer,
             attrs.type)
-    ELIF CUPY_HIP_VERSION > 0:
+    ELIF 0 < CUPY_HIP_VERSION < 600:
         return PointerAttributes(
             attrs.device,
             <intptr_t>attrs.devicePointer,
             <intptr_t>attrs.hostPointer,
             attrs.memoryType)
+    ELIF CUPY_HIP_VERSION >= 600:
+        return PointerAttributes(
+            attrs.device,
+            <intptr_t>attrs.devicePointer,
+            <intptr_t>attrs.hostPointer,
+            attrs.type)
     ELSE:  # for RTD
         return None
 
diff --git a/cupy_backends/hip/cupy_hip_runtime.h b/cupy_backends/hip/cupy_hip_runtime.h
index 3a1e4efd910..3c1cf5c404c 100644
--- a/cupy_backends/hip/cupy_hip_runtime.h
+++ b/cupy_backends/hip/cupy_hip_runtime.h
@@ -270,12 +270,12 @@ cudaError_t cudaPointerGetAttributes(cudaPointerAttributes *attributes,
                                      const void* ptr) {
     cudaError_t status = hipPointerGetAttributes(attributes, ptr);
     if (status == cudaSuccess) {
-        switch (attributes->memoryType) {
+        switch (attributes->type) {
             case 0 /* hipMemoryTypeHost */:
-                attributes->memoryType = (hipMemoryType)1; /* cudaMemoryTypeHost */
+                attributes->type = (hipMemoryType)1; /* cudaMemoryTypeHost */
                 return status;
             case 1 /* hipMemoryTypeDevice */:
-                attributes->memoryType = (hipMemoryType)2; /* cudaMemoryTypeDevice */
+                attributes->type = (hipMemoryType)2; /* cudaMemoryTypeDevice */
                 return status;
             default:
                 /* we don't care the rest of possibilities */
diff --git a/install/cupy_builder/install_build.py b/install/cupy_builder/install_build.py
index 720917e0fc3..55f15ded2ec 100644
--- a/install/cupy_builder/install_build.py
+++ b/install/cupy_builder/install_build.py
@@ -155,6 +155,11 @@ def get_compiler_setting(ctx: Context, use_hip):
         include_dirs.append(os.path.join(rocm_path, 'include', 'rocrand'))
         include_dirs.append(os.path.join(rocm_path, 'include', 'hiprand'))
         include_dirs.append(os.path.join(rocm_path, 'include', 'roctracer'))
+        include_dirs.append(os.path.join(rocm_path, 'include', 'hipblas'))
+        include_dirs.append(os.path.join(rocm_path, 'include', 'hipsparse'))
+        include_dirs.append(os.path.join(rocm_path, 'include', 'hipfft'))
+        include_dirs.append(os.path.join(rocm_path, 'include', 'rocsolver'))
+        include_dirs.append(os.path.join(rocm_path, 'include', 'rccl'))
         library_dirs.append(os.path.join(rocm_path, 'lib'))
 
     if use_hip:

From 8a6754773c21338c1cef770ca73af27fa50ef88a Mon Sep 17 00:00:00 2001
From: Nakano Azusa <31366253+Azusachan@users.noreply.github.com>
Date: Fri, 3 May 2024 01:05:13 +0800
Subject: [PATCH 2/2] add macro for cupy_hip_runtime.h

---
 cupy_backends/hip/cupy_hip_runtime.h | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/cupy_backends/hip/cupy_hip_runtime.h b/cupy_backends/hip/cupy_hip_runtime.h
index 3c1cf5c404c..3b76357ce60 100644
--- a/cupy_backends/hip/cupy_hip_runtime.h
+++ b/cupy_backends/hip/cupy_hip_runtime.h
@@ -270,6 +270,7 @@ cudaError_t cudaPointerGetAttributes(cudaPointerAttributes *attributes,
                                      const void* ptr) {
     cudaError_t status = hipPointerGetAttributes(attributes, ptr);
     if (status == cudaSuccess) {
+#if HIP_VERSION >= 600
         switch (attributes->type) {
             case 0 /* hipMemoryTypeHost */:
                 attributes->type = (hipMemoryType)1; /* cudaMemoryTypeHost */
@@ -277,6 +278,15 @@ cudaError_t cudaPointerGetAttributes(cudaPointerAttributes *attributes,
             case 1 /* hipMemoryTypeDevice */:
                 attributes->type = (hipMemoryType)2; /* cudaMemoryTypeDevice */
                 return status;
+#elif
+       switch (attributes->memoryType) {
+            case 0 /* hipMemoryTypeHost */:
+                attributes->memoryType = (hipMemoryType)1; /* cudaMemoryTypeHost */
+                return status;
+            case 1 /* hipMemoryTypeDevice */:
+                attributes->memoryType = (hipMemoryType)2; /* cudaMemoryTypeDevice */
+                return status;
+#endif
             default:
                 /* we don't care the rest of possibilities */
                 return status;
From 2ff608e5b026f89a4edb25aa7b4a38f8fdb25a4f Mon Sep 17 00:00:00 2001
From: Selene <lixueying@mail.bnu.edu.cn>
Date: Thu, 23 May 2024 20:41:14 +0800
Subject: [PATCH] Add ROCm 6 in runtime_version

---
 cupy_backends/cuda/libs/_cnvrtc.pxi | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/cupy_backends/cuda/libs/_cnvrtc.pxi b/cupy_backends/cuda/libs/_cnvrtc.pxi
index 9f02b5522..b2b06aa4f 100644
--- a/cupy_backends/cuda/libs/_cnvrtc.pxi
+++ b/cupy_backends/cuda/libs/_cnvrtc.pxi
@@ -139,5 +139,8 @@ cdef SoftLink _get_softlink():
         elif runtime_version < 6_00_00000:
             # ROCm 5.x
             libname = 'libamdhip64.so.5'
+        elif runtime_version < 7_00_00000:
+            # ROCm 7.x
+            libname = 'libamdhip64.so.6'
 
     return SoftLink(libname, prefix, mandatory=True)
-- 
2.44.0

